kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -110,18 +117,57 @@ private CompletableFuture<KeyContext> makeKeyContext(@NonNull K kekId) {
                 }).toCompletableFuture();
     }
 
-    @NonNull
     @Override
+    @NonNull
     @SuppressWarnings("java:S2445")
-    public CompletionStage<Void> encrypt(@NonNull String topicName,
-                                         int partition,
-                                         @NonNull EncryptionScheme<K> encryptionScheme,
-                                         @NonNull List<? extends Record> records,
-                                         @NonNull Receiver receiver) {
-        if (records.isEmpty()) {
-            return CompletableFuture.completedFuture(null);
+    public CompletionStage<MemoryRecords> encrypt(@NonNull String topicName,
+                                                  int partition,
+                                                  @NonNull EncryptionScheme<K> encryptionScheme,
+                                                  @NonNull MemoryRecords records,
+                                                  @NonNull Function<Integer, ByteBufferOutputStream> bufferAllocator) {
+        if (records.firstBatch() == null) {
+            return CompletableFuture.completedFuture(records);
+        }
+        MemoryRecordsBuilder builder = recordsBuilder(allocateBufferForEncode(records, bufferAllocator), records);
+        var encryptionRequests = recordStream(records).toList();
+        CompletionStage<Void> result;
+        if (encryptionRequests.isEmpty()) {
+            result = CompletableFuture.completedFuture(null);
+        }
+        else {
+            result = attemptEncrypt(topicName, partition, encryptionScheme, encryptionRequests, (kafkaRecord, plaintextBuffer, headers) -> {
+                builder.appendWithOffset(kafkaRecord.offset(), kafkaRecord.timestamp(), kafkaRecord.key(), plaintextBuffer, headers);
+            }, 0);
         }
-        return attemptEncrypt(topicName, partition, encryptionScheme, records, receiver, 0);
+        return result.thenApply(unused -> builder.build());
+    }
+
+    @NonNull
+    private static Stream<Record> recordStream(MemoryRecords memoryRecords) {
+        return StreamSupport.stream(memoryRecords.records().spliterator(), false);
+    }
+
+    private static MemoryRecordsBuilder recordsBuilder(@NonNull ByteBufferOutputStream buffer, @NonNull MemoryRecords records) {
```

Comment:
note this is just pulling the existing code into the KeyManager, the MemoryRecords produced will not have the same batching as the input. That can be done later.